docs/DeveloperGuide.md

```diff
@@ -610,6 +610,53 @@ CompletableFuture<List<Application>> applications =
     );
 ```
 
+## Keep a Language Server alive with a Lease
+
+If you need a reference to your language server and need to execute operations with it, 
+for an extended period of time, you can use `LanguageServerItem.keepAlive()` method to create
+a `Lease` on the item. The lease represents a request for lsp4ij not to terminate the language 
+server until the `Lease` is disposed.
+
+```java
+Project project = ...
+CompletableFuture<Lease<LanguageServerItem>> serverLease = 
+        LanguageServerManager.getInstance(project).getLanguageServer("myLanguageServerId")
+        .thenApply(item -> item.keepAlive());
+```
+
+As long as the lease has not been disposed this will request lsp4ij not arbitrarily terminate the language
+server under 'normal circumstances'. This means that the server will not be terminated for example when
+there are no more open editors.
+
+To get access to your language server from the lease call the `get()` method on the lease.
+This will return a LanuageServerItem that is guaranteed to have an 'active' server at that time.
+
+```java
+serverLease.thenAccept(lease -> {
+    // Fetch list of 'applications' using some custom protocol on 'MyLanguageServer'
+    List<Application> applications = ((MyLanguageServer)lease.get().getServer()).getApplications();   
+    ... do something with the result... 
+});
+```
+
+Note that it is still possible under some circumstance that the server might get terminated despite having 
+active leases: 
+
+  - the server process could crash unexpectedly. 
+  - a user could stop the language server from lsp4ij UI.

```

Comment:
I think you should add a link to the LSPApi.md to speak about canCancelServerByUser

Reply:
About that, I played with this api a bit but it doesn't actually stop the user from stopping the server (there appears to be two actions in the ui. One is called 'Pause' and the other 'Stop'. When we use 'Pause' it stops the server. And when we use 'Stop' it stops the server and disables it so it can no longer be restarted.

The api only stops the user from using 'stop' action but not the 'pause' action. So it means even if I override this to return 'false' the user is in fact still able to stop the server via 'pause'.

I'm not sure that is intentional. Perhaps the wording 'pause' and 'stop' are what confused me a bit. As I could not imagine a mechanism to really 'pause' the process, and in fact it doesn't. It effectively terminates the process.

Reply:
I added pointer in the docs. But I still wonder now if this option perhaps should disable both 'pause' and 'stop', since both of these can cause a server process to be terminated. If you think its important to be able to disable one but not the other maybe they should instead be controlled by two separate flags.

Reply:
All in all though I think its probably best for a client/user like myself not to mess with these options. If a user decided to stop a language server they may have very good reason to do so (maybe the server misbehaves somehow or uses too much of the users machine/resources). In that case we shouldn't take away control from the user. So I think its better if client code should instead deal with this some reasonable way. (I.e. in my case I imagine we either just show an error message in the view or disable the view completely if server has been 'disabled/stopped' by the user from the get go.

I.e. I think the ability for a user to stop a server is not something I really expect them to be using under 'normal circumstances'. But if they want do it anyway... then they may have a very good reason for this, and we shouldn't really stop them from doing that.

Reply:
> I added pointer in the docs. But I still wonder now if this option perhaps should disable both 'pause' and 'stop', since both of these can cause a server process to be terminated. If you think its important to be able to disable one but not the other maybe they should instead be controlled by two separate flags.

This Pause action is only available in dev mode. When user will install LSP4IJ, the Pause action will never appear.

I think it can be nice to have this Pause action while your are developping LSP4IJ to emulate for instance crash of language server.

Reply:
>  I think it can be nice to have this Pause action while your are developing...

Okay I understand now and I agree. I didn't realize the only reason I could see the action is because I am running it in 'development mode'.